akka-stream/src/main/scala/akka/stream/impl/io/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.channels.FileChannel
+
+import akka.actor.{ ActorLogging, DeadLetterSuppression, Props }
+import akka.io.DirectByteBufferPool
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+
+import scala.annotation.tailrec
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object SynchronousFilePublisher {
+  def props(f: File, completionPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, completionPromise, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, bytesReadPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
+
+  private var raf: RandomAccessFile = _
+  private var chan: FileChannel = _
+
+  override def preStart() = {
+    try {
+      raf = new RandomAccessFile(f, "r") // best way to express this in JDK6, OpenOption are available since JDK7
+      chan = raf.getChannel
+      readAndSignal(initialBuffer)
```

Comment:
If just "start reading to see if can complete" is the need than one can always use a `.buffer(1, backpressure)` after such publisher. The general idea of the file publisher is to always have a bit of data loaded, before `Request(n)` comes in, so two birds with one stone. It's less interesting in this impl because of the synchronous reads... I was implementing an AsynchronousFileChannel based impl where this gets a bit more interesting (though not necessarily faster).